### PR TITLE
RSA: correct digestinfo_ripemd160_der[]

### LINKED ISSUE #13559: RSA signatures with RIPEMD160 MAC incompatible between openssl-3.0.0-alpha9 and openssl-1.x.x
I am one of the developers of an open source toolkit named DCMTK (www.dcmtk.org) that implements the DICOM standard ("Digital Imaging and Communications in Medicine"), which defines a very common file format and network communication protocol for medical images. DCMTK uses OpenSSL to implement encrypted network connections (TLS) and digital signatures.

I tested the current version of DCMTK, which is know to work correctly with OpenSSL 1.0.1, 1.0.2, 1.1.0 and 1.1.1, with the latest Alpha 9 release of OpenSSL 3.0.0 and noticed that most digital signatures created with previous OpenSSL versions cannot be verified anymore with the same tool (same source code) compiled and linked against OpenSSL 3.0.0-alpha9.

Specifically, this affects RSA signatures using a RIPEMD160 message digest, which is the default for signatures in the DICOM standard. The reason is apparently the PKCS1 padding, implemented in encode_pkcs1() in crypto/rsa/rsa_sign.c, which produces a different byte stream in openssl-3.0.0-alpha9 than in previous releases.

In previous releases (specifically 1.1.0), encode_pkcs1() internally calls i2d_X509_SIG() and produces the following padded byte sequence for a RIPEMD hash:
```
30 20 30 08 06 04 2b 24 03 03 01 02 05 00 04 14                # PKCS#1 padding
eb 64 e6 fe e9 59 7e 6c ab 45 8d 5f ea cc 39 51 05 49 0d 8d    # RIPEMD160 hash
```
In openssl-3.0.0-alpha9, the padding is defined in digestinfo_ripemd160_der[] and produces the folowing byte sequence:
```
30 21 30 09 06 05 2b 24 03 02 01 05 00 04 14                   # PKCS#1 padding
eb 64 e6 fe e9 59 7e 6c ab 45 8d 5f ea cc 39 51 05 49 0d 8d    # RIPEMD160 hash
```

This causes signatures created with OpenSSL 1.x.x to fail verification when using OpenSSL 3.0.0, and vice versa.

The bug can be reproduced by compiling DCMTK with OpenSSL 3.0.0 and verifying a signature created with an earlier version (e.g. using the binaries available from www.dcmtk.org).

### PR DESCRIPTION
**Pull request by levitte**
A couple of numbers were incorrect.

Fixes #13559

### CODE REVIEW

File: crypto/rsa/rsa_sign.c [hunk 1]
```diff
@@ -116,9 +116,9 @@ static const unsigned char digestinfo_mdc2_der[] = {
 # ifndef OPENSSL_NO_RMD160
 /* RIPEMD160 (1 3 36 3 3 1 2) */
 static const unsigned char digestinfo_ripemd160_der[] = {
-    ASN1_SEQUENCE, 0x0c + RIPEMD160_DIGEST_LENGTH,
+    ASN1_SEQUENCE, 0x0e + RIPEMD160_DIGEST_LENGTH,
       ASN1_SEQUENCE, 0x08,
```
Comment (slontis) [inline crypto/rsa/rsa_sign.c:120]: This still is not right..

Comment (slontis) [reply crypto/rsa/rsa_sign.c:120]: ASN1_SEQUENCE, 0x08, is incorrect I think..

Comment (levitte) [reply crypto/rsa/rsa_sign.c]: Fixed

**Comment by levitte** [comment]
It turned out that not only the encoding was incorrect, but the OID used was as well.

**Review by slontis — COMMENTED** [review]
I have a few concerns about this change..

openssl 1.1.1 command line app
./apps/openssl dgst -sigopt rsa_padding_mode:pkcs1 -ripemd160WithRSA -sign rsa_priv.pem -hex in.txt

seems to be producing the oid that it was before...

**Comment by slontis** [comment]
> seems to be producing the oid that it was before...

looks like you changed the oid..

1.1.1 is
06 05 2B 24 03 02 01
05 00

So I think it is ok now..

Ideally we would spit out some 1.1.1 values and sign them again in 3.0.0..

**Review by slontis — APPROVED** [review]

**Comment by levitte** [comment]
Merged

172daa7fc7d8cdf37a5203cfeb3e3030d15c1a12 RSA: correct digestinfo_ripemd160_der[]

**Pull request closed without merging**